### PR TITLE
migrate in samtools, kraken, vcflib

### PR DESCRIPTION
**Pull request by martenson**
part of https://github.com/galaxyproject/tools-devteam/issues/406
xref: https://github.com/galaxyproject/tools-iuc/pull/1060

this is an effort to address part of https://github.com/galaxyproject/tools-devteam/issues/416

### CODE REVIEW

File: tool_collections/samtools/bam_to_sam/bam_to_sam.xml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        </param>
+    </inputs>
+    <outputs>
+        <data format="sam" label="${tool.name} on ${on_string}: converted SAM" name="output1" />
+    </outputs>
+    <tests>
+        <test>
+            <param ftype="bam" name="input1" value="bam_to_sam_in1.bam" />
+            <param name="header" value="-h" />
+            <output file="bam_to_sam_out1.sam" name="output1" sorted="True" />
```
Comment (martenson) [inline tool_collections/samtools/bam_to_sam/bam_to_sam.xml]: should this be just `sort`?
source https://docs.galaxyproject.org/en/latest/dev/schema.html#id74

File: tool_collections/samtools/samtools_stats/samtools_stats.xml [hunk 2]
```diff
@@ -0,0 +1,296 @@
+<tool id="samtools_stats" name="Stats" version="2.0">
```
Comment (martenson) [inline tool_collections/samtools/samtools_stats/samtools_stats.xml]: this file ha `.. ERROR: Invalid tmpmXYONH found. Errors [/tmp/tmpmXYONH:218:0:ERROR:SCHEMASV:SCHEMAV_ELEMENT_CONTENT: Element 'filter': This element is not expected.]
` problem

File: tool_collections/samtools/samtools_stats/samtools_stats.xml [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            <when value="yes">
+                <conditional name="reference_source">
+                    <param name="reference_source_selector" type="select" label="Choose a reference sequence for GC depth">
+                        <option value="cached">Locally cached</option>
+                        <option value="history">History</option>
+                    </param>
+                    <when value="cached">
+                        <param name="ref_file" type="select" label="Using genome">
+                            <options from_data_table="fasta_indexes" />
+                            <filter type="data_meta" ref="input_file" key="dbkey" column="1" />
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_stats/samtools_stats.xml]: This should be:
```
                            <options from_data_table="fasta_indexes">
                                <filter type="data_meta" ref="input_file" key="dbkey" column="1" />
                            </options>
```

File: tool_collections/samtools/samtools_stats/samtools_stats.xml (hunk 2 above)
Comment (nsoranzo) [inline tool_collections/samtools/samtools_stats/samtools_stats.xml]: @martenson See my comment below.

File: tool_collections/samtools/samtools_stats/samtools_stats.xml (hunk 3 above)
Comment (martenson) [inline tool_collections/samtools/samtools_stats/samtools_stats.xml]: thought so, thanks for clarification, fixing

**Comment by martenson** [comment]
both samtools_mpileup tests fail with
```
======================================================================
FAIL: MPileup ( samtools_mpileup ) > Test-2
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/tmpW4VvWu/galaxy-dev/test/functional/test_toolbox.py", line 302, in test_tool
    self.do_it( td )
  File "/tmp/tmpW4VvWu/galaxy-dev/test/functional/test_toolbox.py", line 78, in do_it
    raise e
JobOutputsError: [Errno 2] No such file or directory: '/home/travis/build/galaxyproject/tools-iuc/tool_collections/samtools/samtools_mpileup/test-data/samtools_mpileup_out_2.log'
```

**Comment by bgruening** [comment]
@martenson can we remove all tool_dependencies.xml files and make sure all are condafied?

**Comment by martenson** [comment]
@bgruening will do

File: tool_collections/kraken/macros.xml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0"?>
+<macros>
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="0.10.6_eaf8fb68">kraken</requirement>
+            <requirement type="package" version="0.10.6-eaf8fb68">kraken</requirement>
```
Comment (martenson) [inline tool_collections/kraken/macros.xml]: @davebx Do you know why is the requirement here twice? Once with dash and once with underscore.

Comment (martenson) [inline tool_collections/kraken/macros.xml]: I guess it is because of the bioconda vs TS package names.

**Comment by martenson** [comment]
thank you @yhoogstrate !!

**Comment by bgruening** [comment]
it seems there is a wrong samtools version used: 
```
[main] unrecognized command '--version'
[sam_header_read2] 1 sequences loaded.
sort: invalid option -- 'O'
sort: invalid option -- 'T'
```

**Comment by yhoogstrate** [comment]
This error is comming from samtools 0.1.19, installed by travis: https://github.com/galaxyproject/tools-iuc/blob/master/.travis.yml#L20 . I think Galaxy does not override it with 1.2.* properly and uses 0.1,19 that was/is necessary for the Galaxy backend, instead. 

During planemo tests samtools 1.3.1 and 1.2.* are being installed. As 1.2 is the dependency of the tool wrapper, I suppose 1.3.1 is now provided by Galaxy to avoid manually installing 0.1.19, making the installation in the travis config file is obsolete? Would removing the samtools dependency from travis be a good idea?

**Comment by yhoogstrate** [comment]
Could the failure be due to `GALAXY_PYTHON='which python'`, which may pick the python provided by conda?

File: tool_collections/kraken/kraken/kraken.xml [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+        #end if
+
+        #if $single_paired.single_paired_selector == "yes":
+            #if $forward_input.is_of_type( 'fastq' ):
+                --fastq-input
+            #else:
+                --fasta-input
+            #end if
+            "$forward_input" "$reverse_input"
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken/kraken.xml]: This line should be:
```
            "${single_paired.forward_input}" "${single_paired.reverse_input}"
```

File: tool_collections/kraken/kraken/kraken.xml [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+            #end if
+            "${single_paired.input_pair.forward}" "${single_paired.input_pair.reverse}"
+            ${single_paired.check_names}
+        #else:
+            #if $input_sequences.is_of_type( 'fastq' ):
+                --fastq-input
+            #else:
+                --fasta-input
+            #end if
+            "$input_sequences"
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken/kraken.xml]:
```
            '${single_paired.input_sequences}'
```

File: tool_collections/kraken/kraken/kraken.xml [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        #elif $single_paired.single_paired_selector == "collection":
+            #if $single_paired.input_pair.forward.is_of_type( 'fastq' ):
+                --fastq-input
+            #else:
+                --fasta-input
+            #end if
+            "${single_paired.input_pair.forward}" "${single_paired.input_pair.reverse}"
+            ${single_paired.check_names}
+        #else:
+            #if $input_sequences.is_of_type( 'fastq' ):
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken/kraken.xml]:
```
            #if $single_paired.input_sequences.is_of_type('fastq')
```

File: tool_collections/kraken/kraken/kraken.xml [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        
+        ${only_classified_output}
+
+        #if str( $quick_operation.quick ) == "yes":
+            --quick
+            --min-hits ${quick_operation.min_hits}
+
+        #end if
+
+        #if $single_paired.single_paired_selector == "yes":
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken/kraken.xml]:
```
        #if $single_paired.single_paired.single_paired_selector == 'yes'
```

File: tool_collections/kraken/kraken/kraken.xml [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  --fastq-input             Input is FASTQ format
+  --quick                   Quick operation (use first hit or hits)
+  --min-hits NUM            In quick op., number of hits req'd for classification
+                            NOTE: this is ignored if --quick is not specified
+  --unclassified-out        Print unclassified sequences to filename
+  --classified-out          Print classified sequences to filename
+
+  --only-classified-output  Print no Kraken output for unclassified sequences
+                          
+------
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken/kraken.xml]: All options are already documented inside the `<inputs>` section, duplicating here only complicates the wrapper maintenance.

File: tool_collections/kraken/kraken_report/kraken-mpa-report.xml [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+-----
+
+**What is Does**
+
+Kraken-mpa-report summarizes read counts across taxonomic ranks for multiple samples. This is convenient for comparing results across multiple experiments, conditions, locations, etc. It supports the following parameters::
+
+--show-zeros    Display taxa even if they lack a read in any sample
+--header-line   Display a header line indicating sample IDs
+                (sample IDs are the filenames)
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken_report/kraken-mpa-report.xml]: Remove option documentation.

File: tool_collections/kraken/macros.xml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+            <exit_code description="Tool exception" level="fatal" range="1:" />
+        </stdio>
+    </xml>
+    <xml name="version_command">
+        <version_command>export LC_ALL=C &amp;&amp; kraken --version</version_command>
+    </xml>
+    <xml name="input_database">
+        <param label="Select a Kraken database" name="kraken_database" type="select">
+            <options from_data_table="kraken_databases">
+                <validator message="No databases are available built-in" type="no_options" />
```
Comment (nsoranzo) [inline tool_collections/kraken/macros.xml]: "No databases are available built-in" -> "No Kraken database is available"

File: tool_collections/samtools/bam_to_cram/samtools_bam_to_cram.xml [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+            <output name="output_alignment" file="test.cram"  compare="sim_size" delta="250" />
+        </test>
+    </tests>
+
+    <help>
+
+**What this tool does**
+
+Converts alignments from the BAM format to the CRAM format using the SAMTools_ toolkit. The CRAM format does additional compression relative to the reference genome which makes the compression in terms of file size more effictient.
```
Comment (nsoranzo) [inline tool_collections/samtools/bam_to_cram/samtools_bam_to_cram.xml]: s/effictient/efficient/

File: tool_collections/samtools/bam_to_cram/.shed.yml [hunk 13]
```diff
@@ -0,0 +1,13 @@
+categories:
+- Convert Formats
+- SAM
+description: Converts alignments in BAM format to CRAM format.
+long_description: |
+  This tool uses the SAMtools toolkit to convert BAM to CRAM files. The CRAM
+  format does additional compression relative to the reference genome which
+  makes the compression more effictient in terms of file size.
```
Comment (nsoranzo) [inline tool_collections/samtools/bam_to_cram/.shed.yml]: s/effictient/efficient/

File: tool_collections/samtools/sam_to_bam/sam_to_bam.xml [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+        </test>
+    </tests>
+    <help>
+<![CDATA[
+**What it does**
+
+Converts SAM dataset into its binary, BAM, representation using ``samtools sort`` and ``view`` commands::
+
+  samtools sort -O bam -o sorted_input.bam [INPUT SAM] 
+  samtools view -b -h -o -T [REFERENCE GENOME] [OUTPUT BAM] sorted_input.bam
```
Comment (nsoranzo) [inline tool_collections/samtools/sam_to_bam/sam_to_bam.xml]: This is not what the tool does.

File: tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    <expand macro="stdio"></expand>
+    <expand macro="version_command"></expand>
+    <command><![CDATA[
+        for bamfile in 
+        #for dataset in $input_bams:
+            "${dataset}"
+        #end for
+        ; do
+            if [ ! -f \$bamfile.bai ] ; then
+                ln -s \$bamfile && samtools index `basename \$bamfile` ;
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml]:
```bash
                ln -s "\$bamfile" && samtools index "$(basename "\$bamfile")";
```

Here double-quotes are the only option.

File: tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+            else
+                ln -s \$bamfile && ln -s \$bamfile.bai ;
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml]:
```bash
                ln -s "\$bamfile" && ln -s "\$bamfile.bai";
```

File: tool_collections/samtools/sam_to_bam/sam_to_bam.xml (hunk 14 above)
Comment (martenson) [inline tool_collections/samtools/sam_to_bam/sam_to_bam.xml]: @nsoranzo and can you please describe what it does?

File: tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml [hunk 17]
```diff
@@ -7,22 +7,18 @@
...
-        done ;
-        samtools bedcov "${input_bed}"
-        #for dataset in $input_bams:
-            `basename "${dataset}"`
-        #end for
-        > "${output}"
+        
+        samtools bedcov
+            '${input_bed}'
+            #for $i, $bam in enumerate( $input_bams ):
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml]: Alternatively:
```
#for $i in range(len($input_bams))
```

File: tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml [hunk 18]
```diff
@@ -15,7 +15,7 @@
         
         samtools bedcov
             '${input_bed}'
-            #for $i, $bam in enumerate( $input_bams ):
+            #for $i in range(len( $input_bams )):
```
Comment (bgruening) [inline tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml]: I liked it more before, but leave it like it is. `range` has a different behavior on python2 and python3 and enumerate is exactly for these kind of things imho.

Comment (yhoogstrate) [inline tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml]: I think this concern is valid since `xrange` is no part of python3 but replaces `range`, while in python 2 `range` may cause terrible performance with large lists.

Comment (bgruening) [inline tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml]: Cheetah is python2 only at the moment (when do we fix this? ;)), but I encourage people to use enumerate where I can, maybe this is just a personal opinion.

Comment (peterjc) [inline tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml]: I suspect we'd hit command line length limits and/or maximum open file limits here before before Python2's ``range`` vs ``xrange`` performance became an issue.

Comment (nsoranzo) [inline tool_collections/samtools/samtools_bedcov/samtools_bedcov.xml]: Quoting from https://docs.python.org/2/library/functions.html#xrange :
The advantage of `xrange()` over `range()` is minimal (since `xrange()` still has to create the values when asked for them) except when a very large range is used on a memory-starved machine or when all of the range’s elements are never used (such as when the loop is usually terminated with `break`)

File: tool_collections/kraken/kraken_filter/kraken-filter.xml [hunk 19]
```diff
@@ -8,15 +8,18 @@
...
+        
+        kraken-filter
+            @INPUT_DATABASE@
+            --threshold $threshold '${input}' >
+            '$filtered_output'
+    ]]></command>
     <inputs>
         <param format="tabular" label="Kraken output" name="input" type="data" help="Select taxonomy classification produced by kraken"/>
-        <param label="Confidence threshold" max="1" min="0" name="threshold" type="float" value="0" help="--threshold; A number between 0 and 1; default=0"/>
+        <param max="1" min="0" name="threshold" type="float" value="0" label="Confidence threshold" help="A number between 0 and 1; default=0" argument="--threshold"/>
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken_filter/kraken-filter.xml]: You don't need `name` if you use the `argument` attribute. And I would put `argument` first.

**Comment by nsoranzo** [comment]
@yhoogstrate Amazing work!

**Comment by yhoogstrate** [comment]
@nsoranzo Am I correct that using arguments that way requires settings `profile='xx.xx'`, and do you know in which galaxy version `--argument` was implemented, `16.04`??

**Comment by yhoogstrate** [comment]
The samtools warppers use samtools==1.2 but the bioconda build is old and is incorrectly linked to `ncurses`. The 1.2 recepie has been removed but I could re-insert it and rebuild it. What is the preferred solution here, staying with samtools 1.2 or upgrading to 1.3?

**Comment by bgruening** [comment]
I think fixing 1.2 is worth it.

**Comment by mvdbeek** [comment]
I also think restoring 1.2 is worth it, but for our wrappers we should be maintaining them properly and migrating to the latest samtool 1.3.1 if possible, since there were lots of improvements over 1.2.

**Comment by mvdbeek** [comment]
In fact a pyup.io - like service would be absolutely fantastic ... a new version of a recipe in bioconda would automatically trigger a PR on all galaxy tools that use that package ... :).

**Comment by bgruening** [comment]
@mvdbeek indeed!. We have some ideas here to map `arguments` automatically to the tool parameters to automatically detect new parameters or deleted ones.

File: tool_collections/kraken/kraken/kraken.xml [hunk 20]
```diff
@@ -65,46 +65,48 @@
...
-                <!-- Do absolutely nothing -->
+                <param name="min_hits" argument="--min-hits" type="integer" value="1" label="Number of hits required for classification"
+                       help="min-hits will allow you to require multiple hits before declaring a sequence classified, which can be especially useful with custom databases when testing to see if sequences either do or do not belong to a particular genome; default=1"/>
             </when>
+            <when value="no"/><!-- Do absolutely nothing -->
         </conditional>
 
-        <param name="only_classified_output" type="boolean" checked="False" truevalue="--only-classified-output" falsevalue="" label="Print no Kraken output for unclassified sequences" help="--only-classified-output"/>
+        <param name="only_classified_output" argument="--only-classified-output" type="boolean" checked="False"
+               truevalue="--only-classified-output" falsevalue=""
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken/kraken.xml]: I'm not a big fan of splitting xml elements on multiple lines, it makes it more difficult to grep for a combination of element/attributes. I think it is ok to have `label` and `help` on separate lines though.

No big deal anyway, just wanted to mention it.

File: tool_collections/samtools/samtools_rmdup/samtools_rmdup.xml [hunk 21]
```diff
@@ -1,11 +1,8 @@
-<tool id="samtools_rmdup" name="RmDup" version="2.0">
+<tool id="samtools_rmdup" name="RmDup" version="2.1">
     <description>remove PCR duplicates</description>
     <macros>
         <import>macros.xml</import>
     </macros>
-    <requirements>
-        <requirement type="package" version="0.1.19">samtools</requirement>
-    </requirements>
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_rmdup/samtools_rmdup.xml]: @yhoogstrate Why did you remove the requirement?

Comment (yhoogstrate) [inline tool_collections/samtools/samtools_rmdup/samtools_rmdup.xml]: I did that to make it consistent with the other samtools wrappers and because rmdup is also part of `1.2.0`. Are you aware of a (historical) reason to hold the dependency back for just this wrapper?

Comment (nsoranzo) [inline tool_collections/samtools/samtools_rmdup/samtools_rmdup.xml]: I mean that you removed the requirement altogether, I guess you should add
```
<expand macro="requirements" />
```

File: tool_collections/samtools/samtools_stats/samtools_stats.xml [hunk 22]
```diff
@@ -192,7 +197,7 @@
             <when value="nofilter" />
 
         </conditional>
-        <param name="gc_depth" type="float" value="20000" label="GC-depth bin size" help="--GC-depth; decreasing bin size increases memory requirement; default = 20000.0"/>
+        <param name="gc_depth" argument="--GC-depth" type="float" value="20000" label="Bin size GC-depth" help="Decreasing bin size increases memory requirement; default = 20000.0"/>
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_stats/samtools_stats.xml]: Label should be: `Size of GC-depth bins`

File: tool_collections/kraken/kraken/.shed.yml [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+  Kraken is a system for assigning taxonomic labels to short DNA
+  sequences, usually obtained through metagenomic studies. Previous attempts by other
+  bioinformatics software to accomplish this task have often used sequence alignment
+  or machine learning techniques that were quite slow, leading to the development
+  of less sensitive but much faster abundance estimation programs. Kraken aims to
+  achieve high sensitivity and high speed by utilizing exact alignments of k-mers
+  and a novel classification algorithm.
+name: kraken
+owner: devteam
+remote_repository_url: https://github.com/galaxyproject/tools-iuc/blob/master/tool_collections/kraken/kraken/
```
Comment (yhoogstrate) [inline tool_collections/kraken/kraken/.shed.yml]: @martenson The kraken wrappers are under devteam and the others under IUC. Is it ok to choose either devteam or IUC for all of them?

Comment (martenson) [inline tool_collections/kraken/kraken/.shed.yml]: @yhoogstrate I do not understand the question. Please rephrase.

Comment (yhoogstrate) [inline tool_collections/kraken/kraken/.shed.yml]: @martenson, oops I think I clicked at the wrong file to explain myself. The following .shed.yml files have their `remote_repository_url:` linked to `tools-devteam`:

```
tool_collections/kraken/kraken_filter/.shed.yml:remote_repository_url: https://github.com/galaxyproject/tools-devteam/blob/master/tool_collections/kraken/kraken_filter/
tool_collections/kraken/kraken_report/.shed.yml:remote_repository_url: https://github.com/galaxyproject/tools-devteam/blob/master/tool_collections/kraken/kraken_report/
tool_collections/kraken/kraken_translate/.shed.yml:remote_repository_url: https://github.com/galaxyproject/tools-devteam/blob/master/tool_collections/kraken/kraken_translate/
```

While all other tools have this linked to `tools-iuc`. Is this on purpose?

Comment (nsoranzo) [inline tool_collections/kraken/kraken/.shed.yml]: I'm pretty sure these are just leftovers, all `remote_repository_url` should link to `tools-iuc` .

Comment (martenson) [inline tool_collections/kraken/kraken/.shed.yml]: @yhoogstrate I think @nsoranzo is right. I fixed those 3.

File: tool_collections/samtools/cram_to_bam/samtools_cram_to_bam.xml [hunk 24]
```diff
@@ -0,0 +1,114 @@
+<tool id="samtools_cram_to_bam" name="samtools CRAM to BAM" version="1.2.0">
+    <description>convert CRAM alignments to BAM format</description>
+
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements"/>
+    <expand macro="version_command"/>
+    <expand macro="stdio"/>
```
Comment (nsoranzo) [inline tool_collections/samtools/cram_to_bam/samtools_cram_to_bam.xml]:
```
<expand macro="stdio"/>
```
should be moved between `requirements` and `version_command`

File: tool_collections/kraken/kraken_filter/kraken-filter.xml [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+    <outputs>
+        <data format="tabular" name="filtered_output" />
+    </outputs>
+    <tests>
+        <test>
+            <param name="input" value="kraken_filter_test1.tab"/>
+            <param name="threshold" value="0"/>
+            <param name="kraken_database" value="test_db"/>
+
+            <output name="output" file="kraken_filter_test1_output.tab" ftype="tabular"/>
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken_filter/kraken-filter.xml]: s/"output"/"filtered_output"/

File: tool_collections/samtools/samtools_rmdup/samtools_rmdup.xml [hunk 26]
```diff
@@ -0,0 +1,61 @@
+<tool id="samtools_rmdup" name="RmDup" version="2.1">
+    <description>remove PCR duplicates</description>
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+
+    <expand macro="stdio"/>
+    <expand macro="version_command"/>
+    <expand macro="requirements" />
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_rmdup/samtools_rmdup.xml]:
```
<expand macro="requirements" />
```
should be moved before `stdio`.

File: tool_collections/samtools/samtools_flagstat/samtools_flagstat.xml [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+    <command><![CDATA[
+        samtools flagstat '$input1' > '$output1'
+    ]]></command>
+
+    <inputs>
+        <param name="input1" type="data" format="bam" label="BAM File to report statistics of" />
+    </inputs>
+
+    <outputs>
+    <data name="output1" format="txt" />
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_flagstat/samtools_flagstat.xml]: Indentation.

File: tool_collections/samtools/samtools_flagstat/samtools_flagstat.xml [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+    </outputs>
+
+    <tests>
+        <test>
+          <param name="input1" value="samtools_flagstat_input1.bam" ftype="bam" />
+
+          <output name="output1" file="samtools_flagstat_out1.txt" />
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_flagstat/samtools_flagstat.xml]: Indentation of the previous 2 lines.

File: tool_collections/samtools/samtools_flagstat/samtools_flagstat.xml [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+  100 + 0 read1
+  100 + 0 read2
+  0 + 0 properly paired (0.00%:nan%)
+  0 + 0 with itself and mate mapped
+  25 + 0 singletons (12.50%:nan%)
+  0 + 0 with mate mapped to a different chr
+  0 + 0 with mate mapped to a different chr (mapQ>=5)
+  
+  </help>
+  <expand macro="citations"/>
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_flagstat/samtools_flagstat.xml]: Indentation of the previous 2 lines.

File: tool_collections/kraken/kraken_filter/kraken-filter.xml [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+    </macros>
+    <expand macro="requirements" />
+    <expand macro="version_command" />
+    <command detect_errors="exit_code"><![CDATA[
+        @SET_DATABASE_PATH@ &&
+        
+        kraken-filter
+            @INPUT_DATABASE@
+            --threshold $threshold
+            '${input}' >
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken_filter/kraken-filter.xml]: I'd move the `>` on the next line.

File: tool_collections/kraken/kraken_report/kraken-report.xml [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements" />
+    <expand macro="version_command" />
+    <command detect_errors="exit_code"><![CDATA[
+        @SET_DATABASE_PATH@ &&
+        
+        kraken-report
+            @INPUT_DATABASE@
+            '${kraken_output}' >
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken_report/kraken-report.xml]: I'd move the `>` on the next line.

File: tool_collections/kraken/kraken_translate/kraken-translate.xml [hunk 32]
```diff
@@ -0,0 +1,61 @@
...
+    </macros>
+    <expand macro="requirements" />
+    <expand macro="version_command" />
+    <command detect_errors="exit_code"><![CDATA[
+        @SET_DATABASE_PATH@ &&
+        
+        kraken-translate
+            @INPUT_DATABASE@
+            $mpa_format
+            '${input}' >
```
Comment (nsoranzo) [inline tool_collections/kraken/kraken_translate/kraken-translate.xml]: I'd move the `>` on the next line.

File: tool_collections/samtools/samtools_reheader/samtools_reheader.xml [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements"/>
+    <expand macro="stdio"/>
+    <expand macro="version_command"/>
+    <command><![CDATA[
+        samtools reheader
+            '${input_header}'
+            '${input_file}' >
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_reheader/samtools_reheader.xml]: I'd move the `>` on the next line.

File: tool_collections/samtools/samtools_rmdup/samtools_rmdup.xml (hunk 26 above)
Comment (nsoranzo) [inline tool_collections/samtools/samtools_rmdup/samtools_rmdup.xml]: Also, rmdup should stay on 0.1.19 or move to 1.3.*, see https://github.com/galaxyproject/tools-devteam/issues/284 and https://github.com/galaxyproject/tools-devteam/issues/286 .

**Comment by martenson** [comment]
@nsoranzo I addressed your comments

**Comment by yhoogstrate** [comment]
Nice @martenson , `planemo test --update_test_data` and I think we're there :+1:

File: tool_collections/samtools/macros.xml [hunk 34]
```diff
@@ -0,0 +1,71 @@
+<macros>
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="1.3">samtools</requirement>
```
Comment (nsoranzo) [inline tool_collections/samtools/macros.xml]: s/1.3/1.3.1/

1.3.1 is the latest and is already used by `tools/htseq_count/htseq-count.xml`

File: tool_collections/samtools/samtools_flagstat/samtools_flagstat.xml [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+    <command><![CDATA[
+        samtools flagstat '$input1' > '$output1'
+    ]]></command>
+
+    <inputs>
+        <param name="input1" type="data" format="bam" label="BAM File to report statistics of" />
+    </inputs>
+
+    <outputs>
+      <data name="output1" format="txt" />
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_flagstat/samtools_flagstat.xml]: Indentation.

File: tool_collections/samtools/macros.xml (hunk 34 above)
Comment (martenson) [inline tool_collections/samtools/macros.xml]: should this go together with a version bump?

File: tool_collections/samtools/samtools_stats/samtools_stats_tool_test_output.html [hunk 36]
```diff
@@ -0,0 +1,372 @@
+<!DOCTYPE html>
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_stats/samtools_stats_tool_test_output.html]: This file should be removed.

File: tool_collections/samtools/macros.xml (hunk 34 above)
Comment (nsoranzo) [inline tool_collections/samtools/macros.xml]: Yes, all `samtools` tool versions should be bumped, for `bam_to_cram`, `cram_to_bam` and `samtools_split` the new version should be `1.3.1` (using a macro ideally), all other tools have already a major version `2` so I'd increase the 'patch' version number, hoping one day we can sync again with the upstream version.

File: tool_collections/samtools/samtools_mpileup/test-data/samtools/mpileup/samtools_mpileup_out_1.pileup [hunk 37]
```diff
@@ -0,0 +1,43 @@
+phiX174	1411	A	1	^P.	$
```
Comment (bgruening) [inline tool_collections/samtools/samtools_mpileup/test-data/samtools/mpileup/samtools_mpileup_out_1.pileup]: Is this file correct here?

File: tool_collections/samtools/samtools_mpileup/test-data/samtools_mpileup_out_1.pileup [hunk 38]
```diff
@@ -0,0 +1,43 @@
+phiX174	1411	A	0				1
+phiX174	1412	G	0				2,1,1
```
Comment (bgruening) [inline tool_collections/samtools/samtools_mpileup/test-data/samtools_mpileup_out_1.pileup]: According to the last failing test, the last column needs to be removed to pass.
Is this correct? @nekrut have you used mpileup and now what the last column is and if it is ok if it get's removed?

Comment (yhoogstrate) [inline tool_collections/samtools/samtools_mpileup/test-data/samtools_mpileup_out_1.pileup]: This looks like a fixed bug in samtools to me and results in backwards incompatibility.
The option `-O` returns the base positions of the corresponding reads.
The 4th column indicates zero bases are included in the mpileup, while the last column indicates multiple reads are present. To me the behaviour in 1.3 looks appropriate.

Comment (bgruening) [inline tool_collections/samtools/samtools_mpileup/test-data/samtools_mpileup_out_1.pileup]: But this file has multiple columns, the end is just a lot of tabulars afaik.

File: tool_collections/samtools/samtools_stats/test-data/samtools_stats_out2/gcc.tab [hunk 39]
```diff
@@ -0,0 +1,252 @@
+# ACGT content per cycle. The columns are: cycle, and A,C,G,T counts (percent)
+1	21.50	29.50	33.50	15.50
```
Comment (nsoranzo) [inline tool_collections/samtools/samtools_stats/test-data/samtools_stats_out2/gcc.tab]: This file also needs to be updated.

**Comment by nsoranzo** [comment]
It's green :tada: , thanks @yhoogstrate!

**Comment by bgruening** [comment]
:tada: !!!

**Comment by bgruening** [comment]
Thanks all, this is awesome!

**Comment by nsoranzo** [comment]
:worried:  samtools tool versions have not been updated for the new samtools 1.3 requirement, see https://github.com/galaxyproject/tools-iuc/pull/1148/files#r104935513

**Comment by nsoranzo** [comment]
Luckily all repo updates have failed because @martenson hasn't given the necessary permissions to `iuc` yet, so we can still fix the tool versions.

**Comment by martenson** [comment]
@nsoranzo yay! :)

**Comment by martenson** [comment]
all repos from this PR have been added to IUC-permitted
